### PR TITLE
fixed: Cache low speed condition check could report wrong rate (Superseded by #20760)

### PR DESCRIPTION
**Pull request by arnova**
Superseded by https://github.com/xbmc/xbmc/pull/20760 to resolve Jenkins error.  Upon merge this PR will be closed.

Previously a boolean flag was used to "remember" whether a low speed read condition occurred in the filecache. The problem with this is that since it's a past event, reporting the current rate is wrong and may be confusing. I've resolved this may changing the boolean into an unsigned which either stores the low speed rate or is reset to 0 when none occurred.

@AlwinEsch : Since it also affects the addon vfs, I suspect "something" needs to be bumped somewhere. Please let me know what.

### CODE REVIEW

File: xbmc/addons/kodi-dev-kit/include/kodi/c-api/filesystem.h [hunk 1]
```diff
@@ -221,7 +221,7 @@ extern "C"
     uint64_t forward;
     unsigned int maxrate;
     unsigned int currate;
-    bool lowspeed;
+    unsigned int lowrate;
```
Comment (AlwinEsch) [inline xbmc/addons/kodi-dev-kit/include/kodi/c-api/filesystem.h:224]: By change here need `ADDON_INSTANCE_VERSION_VFS_MIN` and `ADDON_INSTANCE_VERSION_VFS` a version increase and rebuild of addons.
On this place https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-dev-kit/include/kodi/versions.h#L164-L165 and as 3.0.1 would be enough.

File: xbmc/addons/kodi-dev-kit/include/kodi/Filesystem.h [hunk 2]
```diff
@@ -267,11 +267,11 @@ class ATTRIBUTE_HIDDEN CacheStatus
   /// @brief Get average read rate from source file since last position change.
   unsigned int GetCurrentRate() { return m_cStructure->currate; }
 
-  /// @brief Set cache low speed condition detected.
-  void SetLowspeed(bool lowspeed) { m_cStructure->lowspeed = lowspeed; }
+  /// @brief Set cache low speed rate detected.
```
Comment (AlwinEsch) [inline xbmc/addons/kodi-dev-kit/include/kodi/Filesystem.h:270]: For here then a description in which format the integer values are set.

File: xbmc/filesystem/IFileTypes.h [hunk 3]
```diff
@@ -51,7 +51,7 @@ struct SCacheStatus
   uint64_t forward;  /**< number of bytes cached forward of current position */
   unsigned maxrate;  /**< maximum number of bytes per second cache is allowed to fill */
   unsigned currate;  /**< average read rate from source file since last position change */
-  bool     lowspeed; /**< cache low speed condition detected? */
+  unsigned lowrate;  /**< low speed rate (if any else 0) */
```
Comment (AlwinEsch) [inline xbmc/filesystem/IFileTypes.h:54]: Also here in which format the rate comes. Seems is "bits per second"?

**Review by AlwinEsch — CHANGES_REQUESTED** [review]
To bring is OK and are not so much VFS add-ons used where need update.
Maybe I can look to combine it with my Sandbox changes, or with instance settings way https://github.com/AlwinEsch/kodi/pull/46 (where need on all add-ons version increase).

File: xbmc/addons/kodi-dev-kit/include/kodi/c-api/filesystem.h (hunk 1 above)
Comment (AlwinEsch) [reply xbmc/addons/kodi-dev-kit/include/kodi/c-api/filesystem.h:224]: Overseen, need then also here a increase https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-dev-kit/include/kodi/versions.h#L65-L66 of `ADDON_GLOBAL_VERSION_FILESYSTEM_MIN` and `ADDON_GLOBAL_VERSION_FILESYSTEM`, then much addons :sweat_smile:.

**Comment by arnova** [comment]
@AlwinEsch : Updated. Please check if I did the addon changes correctly.

File: xbmc/filesystem/IFileTypes.h [hunk 4]
```diff
@@ -48,10 +48,10 @@ struct SNativeIoControl
...
 struct SCacheStatus
 {
-  uint64_t forward;  /**< number of bytes cached forward of current position */
-  unsigned maxrate;  /**< maximum number of bytes per second cache is allowed to fill */
-  unsigned currate;  /**< average read rate from source file since last position change */
-  bool     lowspeed; /**< cache low speed condition detected? */
+  uint64_t forward; /**< number of bytes cached forward of current position */
+  unsigned maxrate; /**< maximum allowed read(fill) rate (bytes/second) */
+  unsigned currate; /**< average read rate (bytes/second) since last position change */
+  unsigned lowrate; /**< low speed read rate (bytes/second) (if any, else 0) */
```
Comment (lrusak) [inline xbmc/filesystem/IFileTypes.h:54]: There seems to be a mixture of `unsigned` and `unsigned int` being used. Do you think we could settle on using a sized type instead like `uint32_t`?

**Review by AlwinEsch — APPROVED** [review]
About addons it would be OK now.

**Comment by AlwinEsch** [comment]
I think it might be good to combine with this part https://github.com/AlwinEsch/kodi/pull/46/commits/b625ae1433e2d30e77d48609c038a634f37c26c4 and the C++17. So all addons would only need a single update.

File: xbmc/filesystem/IFileTypes.h (hunk 4 above)
Comment (arnova) [reply xbmc/filesystem/IFileTypes.h:54]: Yes sounds good. Obviously this is all legacy stuff.

**Comment by arnova** [comment]
> I think it might be good to combine with this part [AlwinEsch@b625ae1](https://github.com/AlwinEsch/kodi/commit/b625ae1433e2d30e77d48609c038a634f37c26c4) and the C++17. So all addons would only need a single update.

Sure, how would you like to handle this?

**Comment by AlwinEsch** [comment]
Will bring a request tonight and think https://github.com/xbmc/xbmc/pull/20699 also ready to come. Also then https://github.com/xbmc/xbmc/pull/20508 can comes in and all addons only one time update :smile:

File: xbmc/filesystem/IFileTypes.h (hunk 4 above)
Comment (arnova) [reply xbmc/filesystem/IFileTypes.h]: @lrusak : Added a commit with the unsigned -> uint32_t changes

**Review by lrusak — APPROVED** [review]
LGTM. Thanks for the improvement and the change to use `uint32_t` 👍

**Comment by lrusak** [comment]
jenkins build this please

**Comment by lrusak** [comment]
jenkins build this please

**Comment by AlwinEsch** [comment]
Have fixed the conflicts inside here by github, see last commit. Need to bring in before addon requests becomes merged.

**Comment by AlwinEsch** [comment]
jenkins build this please

**Comment by AlwinEsch** [comment]
jenkins build this please

**Comment by AlwinEsch** [comment]
jenkins build this please

**Comment by phunkyfish** [comment]
jenkins build this please

**Comment by phunkyfish** [comment]
jenkins build this please

**Comment by phunkyfish** [comment]
Superseded by #20760 to resolve Jenkins error. Closing this PR.

**Pull request closed without merging**